lib/rust/profiling/src/frame_stats.rs

```diff
@@ -0,0 +1,236 @@
+//! This module helps to aggregate various per-frame performance statistics, collected over labeled
+//! time intervals.
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+
+use enso_prelude::*;
+
+use enso_logger::DefaultWarningLogger as Logger;
+use enso_logger::*;
+use serde::Deserialize;
+use serde::Serialize;
+
+
+
+// ====================
+// === IntervalsMap ===
+// ====================
+
+type IntervalsMap = HashMap<String, Bundle>;
+
+thread_local! {
+    static ACTIVE_INTERVALS: RefCell<IntervalsMap> = RefCell::new(IntervalsMap::new());
+}
+
+/// Starts a new named time interval, during which frame statistics will be collected.
+pub fn start_interval(label: &str) {
+    let logger = Logger::new("Profiling_Stats");
+    ACTIVE_INTERVALS.with(|intervals| {
+        let found = intervals.borrow_mut().insert(label.to_string(), Bundle::default());
+        if found.is_some() {
+            warning!(logger, "Trying to collect profiling stats for a process with same label as already existing one - values will be skewed for: {label:?}");
+        }
+    });
+}
+
+/// Finishes collecting frame statistics for a specific named interval. Returns aggregate data
+/// collected since the start of the the interval.
+pub fn end_interval(label: &str) -> Option<Bundle> {
+    let logger = Logger::new("Profiling_Stats");
+    ACTIVE_INTERVALS.with(|intervals| {
+        match intervals.borrow_mut().remove(label) {
+            None => {
+                warning!(logger, "Trying to finalize profiling stats for a process with a label not registered before: {label:?}");
+                None
+            },
+            Some(bundle) if bundle.frames_count == 0 => None,
+            Some(bundle) => Some(bundle),
+        }
+    })
+}
+
+/// A snapshot of values of different metrics at a particular frame, together with human-readable descriptions of those metrics.
+pub type LabeledSamples = Vec<(&'static str, f64)>;
```

Comment:
Typically it's better to use more generic and lightweight `&[T]` instead of `&Vec<T>` in function arguments.